src/rdsn/src/common/backup_restore_common.cpp

```diff
@@ -36,5 +44,41 @@ const std::string backup_restore_constant::BACKUP_ID("restore.backup_id");
 const std::string backup_restore_constant::SKIP_BAD_PARTITION("restore.skip_bad_partition");
 const std::string backup_restore_constant::RESTORE_PATH("restore.restore_path");
 
+std::string get_backup_root(const std::string &backup_root,
+                            const std::string &user_defined_root_path)
+{
+    if (user_defined_root_path.empty()) {
+        return backup_root;
+    }
+    return utils::filesystem::path_combine(user_defined_root_path, backup_root);
+}
+
+std::string get_backup_path(const std::string &root,
+                            const std::string &app_name,
+                            const int32_t app_id,
+                            const int64_t backup_id,
+                            const bool is_compatible)
+{
+    std::string str_app = app_name + "_" + std::to_string(app_id);
+    std::stringstream ss;
+    if (!is_compatible) {
+        ss << root << "/" << str_app << "/" << backup_id;
+    } else {
+        ss << root << "/" << backup_id << "/" << str_app;
+    }
+    return ss.str();
```

Comment:
how about use fmt::format instead?

Reply:
Done, and I resolve the following similar conversation.